go/cmd/vtbackup/vtbackup.go

```diff
@@ -14,120 +14,331 @@ See the License for the specific language governing permissions and
 	}
+	// Shut down mysqld when we're done.
+	defer func() {
+		// Be careful not to use the original context, because we don't want to
+		// skip shutdown just because we timed out waiting for other things.
+		ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+		defer cancel()
+		mysqld.Shutdown(ctx, mycnf, false)
+	}()
 
-	topoServer := topo.Open()
-	mysqld := mysqlctl.NewMysqld(dbcfgs)
-	dir := fmt.Sprintf("%v/%v", *initKeyspace, *initShard)
+	extraEnv := map[string]string{
+		"TABLET_ALIAS": topoproto.TabletAliasString(tabletAlias),
+	}
 
-	pos, err := mysqlctl.Restore(context.Background(), mycnf, mysqld, dir, *concurrency, extraEnv, map[string]string{}, logutil.NewConsoleLogger(), true, dbName)
-	switch err {
-	case nil:
-		// Horray Evenything worked
-		// We have restored a backup ( If one existed, now make sure replicatoin is started
-		if err := resetReplication(context.Background(), pos, mysqld); err != nil {
-			log.Fatalf("Error Starting Replication %v", err)
+	// In initial_backup mode, just take a backup of this empty database.
+	if *initialBackup {
+		// Take a backup of this empty DB without restoring anything.
+		// First, initialize it the way InitShardMaster would, so this backup
+		// produces a result that can be used to skip InitShardMaster entirely.
+		// This involves resetting replication (to erase any history) and then
+		// executing a statement to force the creation of a replication position.
+		mysqld.ResetReplication(ctx)
+		cmds := mysqlctl.CreateReparentJournal()
+		if err := mysqld.ExecuteSuperQueryList(ctx, cmds); err != nil {
+			return fmt.Errorf("can't initialize database with reparent journal: %v", err)
 		}
+		// Now we're ready to take the backup.
+		name := backupName(time.Now(), tabletAlias)
+		if err := mysqlctl.Backup(ctx, mycnf, mysqld, logutil.NewConsoleLogger(), backupDir, name, *concurrency, extraEnv); err != nil {
+			return fmt.Errorf("backup failed: %v", err)
+		}
+		log.Info("Initial backup successful.")
+		return nil
+	}
 
+	// Restore from backup.
+	dbName := *initDbNameOverride
+	if dbName == "" {
+		dbName = fmt.Sprintf("vt_%s", *initKeyspace)
+	}
+
+	log.Infof("Restoring latest backup from directory %v", backupDir)
+	restorePos, err := mysqlctl.Restore(ctx, mycnf, mysqld, backupDir, *concurrency, extraEnv, map[string]string{}, logutil.NewConsoleLogger(), true, dbName)
+	switch err {
+	case nil:
+		log.Infof("Successfully restored from backup at replication position %v", restorePos)
 	case mysqlctl.ErrNoBackup:
-		// No-op, starting with empty database.
+		return fmt.Errorf("no backup found; not starting up empty since -initial_backup flag was not enabled")
 	case mysqlctl.ErrExistingDB:
-		// No-op, assuming we've just restarted.  Note the
-		// replication reporter may restart replication at the
-		// next health check if it thinks it should. We do not
-		// alter replication here.
+		return fmt.Errorf("can't run vtbackup because data directory is not empty")
```

Comment:
Would there ever be a time we want to enable restore from a non-empty tablet. I am thinking if someone makes a PV and they just want to always use that PV to not have to copy all the data to it.

Reply:
That might make sense as an additional feature someday. For now, I needed to enforce this invariant to make sure vtbackup is safe, by allowing the assumption that it always starts empty.